src/properties.c

```diff
@@ -0,0 +1,550 @@
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int set_config(struct knot_thing *thing, int fd, char *group_id,
+		      knot_schema schema, knot_config *config)
+{
+	int rc;
+	int aux;
+	int value_type_aux;
+	knot_value_type tmp_value_type;
+	knot_config config_aux;
+
+	memset(&config_aux, 0, sizeof(knot_config));
+
+	value_type_aux = schema.value_type;
+
+	rc = get_lower_limit(fd, group_id, value_type_aux, &tmp_value_type);
+
+	if (rc > 0) {
+		config_aux.event_flags |= KNOT_EVT_FLAG_LOWER_THRESHOLD;
+		assign_limit(value_type_aux, tmp_value_type,
+			     &config_aux.lower_limit);
+	}
+
+	rc = get_upper_limit(fd, group_id, value_type_aux, &tmp_value_type);
+
+	if (rc > 0) {
+		config_aux.event_flags |= KNOT_EVT_FLAG_UPPER_THRESHOLD;
+		assign_limit(value_type_aux, tmp_value_type,
+			     &config_aux.upper_limit);
+	}
+
+	rc = storage_read_key_int(fd, group_id, CONFIG_TIME_SEC, &aux);
+	if (rc > 0) {
+		config_aux.event_flags |= KNOT_EVT_FLAG_TIME;
+		config_aux.time_sec = aux;
+	}
+
+	rc = storage_read_key_int(fd, group_id, CONFIG_CHANGE, &aux);
+	if (rc > 0)
+		config_aux.event_flags |= KNOT_EVT_FLAG_CHANGE;
+
+	rc = knot_config_is_valid(config_aux.event_flags,
+				  schema.value_type,
+				  config_aux.time_sec,
+				  &config_aux.lower_limit,
+				  &config_aux.upper_limit);
+	if (rc)
+		return -EINVAL;
+
+	*config = config_aux;
+
+	return 0;
+}
+
+static int valid_bit_offset(int bit_offset, int value_type)
+{
+	int value_type_mask;
+
+	switch (bit_offset) {
+	case 1:
+		value_type_mask = KNOT_VALUE_TYPE_BOOL;
+		break;
+	case 8:
+		/* KNoT Protocol doesn't have a matching value type */
+	case 16:
+		/* KNoT Protocol doesn't have a matching value type */
+	case 32:
+		value_type_mask = KNOT_VALUE_TYPE_INT | KNOT_VALUE_TYPE_UINT;
+		break;
+	case 64:
+		value_type_mask = KNOT_VALUE_TYPE_INT64 |
+				  KNOT_VALUE_TYPE_UINT64;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	if (!(value_type & value_type_mask))
+		return -EINVAL;
+
+	return 0;
+}
+
+static int set_modbus_source_properties(struct knot_thing *thing,
+					int fd, char *group_id,
+					knot_schema schema,
+					int *reg_addr, int *bit_offset)
+{
+	int rc;
+	int reg_addr_aux;
+	int bit_offset_aux;
+
+	rc = storage_read_key_int(fd, group_id, MODBUS_REG_ADDRESS,
+				  &reg_addr_aux);
+	if (rc <= 0)
+		return -EINVAL;
+
+	rc = storage_read_key_int(fd, group_id, MODBUS_BIT_OFFSET,
+				  &bit_offset_aux);
+	if (rc <= 0)
+		return -EINVAL;
+
+	rc = valid_bit_offset(bit_offset_aux, schema.value_type);
+	if (rc < 0)
+		return -EINVAL;
+
+	*bit_offset = bit_offset_aux;
+	*reg_addr = reg_addr_aux;
+
+	return 0;
+}
+
+static int set_data_items(struct knot_thing *thing, int fd)
+{
+	int rc;
+	int i;
+	char **data_item_group;
+
+	int sensor_id;
+	int reg_addr;
+	int bit_offset;
+	knot_schema schema;
+	knot_config config;
+
+	data_item_group = get_data_item_groups(fd);
+
+	for (i = 0; data_item_group[i] != NULL ; i++) {
+		rc = set_sensor_id(thing, fd, data_item_group[i], i,
```

Comment:
This 'i' parameter is not necessary anymore

Reply:
done